diff -Naur a/SConstruct b/SConstruct
--- a/SConstruct
+++ b/SConstruct
@@ -850,7 +850,7 @@
     toolchain = ["default"]
 
 env = Environment(tools=toolchain+["textfile", "subst", "recursiveInstall", "UnitsInterfaceBuilder", "wix", "gch"],
-                  ENV={"PATH": os.environ["PATH"]},
+                  ENV={'PATH': os.environ['PATH'], 'CCACHE_DIR': os.environ.get('CCACHE_DIR','')},
                   toolchain=toolchain,
                   **extraEnvArgs)
 
