diff -Naur old/include/cantera/base/fmt.h new/include/cantera/base/fmt.h
--- old/include/cantera/base/fmt.h	2017-01-13 06:28:53.000000000 +0300
+++ new/include/cantera/base/fmt.h	2018-03-05 00:40:32.000000000 +0300
@@ -4,6 +4,7 @@
 #if CT_USE_SYSTEM_FMT
   #include "fmt/format.h"
   #include "fmt/ostream.h"
+  #include "fmt/printf.h"
 #else
   #include "cantera/ext/fmt/format.h"
   #include "cantera/ext/fmt/ostream.h"
