diff -Naur old/doc/SConscript new/doc/SConscript
--- old/doc/SConscript	2017-01-13 06:28:53.000000000 +0300
+++ new/doc/SConscript	2018-08-31 01:20:49.000000000 +0300
@@ -130,8 +130,8 @@
                       mglob(env, '#src/cantera/*', 'h', 'cpp'))
 
     env.Alias('doxygen', docs)
-    install('$inst_docdir/doxygen/html',
-            mglob(localenv, '#/build/docs/doxygen/html', 'html', 'svg', 'css', 'png'))
+    install(localenv.RecursiveInstall, '$inst_docdir/doxygen/html',
+            '#/build/docs/doxygen/html', exclude=['\\.map', '\\.md5'])
 
 if localenv['sphinx_docs']:
     localenv['SPHINXBUILD'] = Dir('#build/docs/sphinx')
