diff -Naur old/SConstruct new/SConstruct
--- old/SConstruct	2017-01-13 06:28:53.000000000 +0300
+++ new/SConstruct	2018-03-14 09:44:29.000000000 +0300
@@ -494,6 +494,11 @@
            each argument ('no') in the Fortran libraries.""",
         True),
     EnumVariable(
+        'googletest',
+        """Control whether check the presence of gtest for the performing
+        of code regression testing via googletest package.""",
+        'default', ('default', 'y', 'n')),
+    EnumVariable(
         'system_googletest',
         """Select whether to use gtest from system installation ('y'), from a
            Git submodule ('n'), or to decide automatically ('default').""",
@@ -824,28 +829,29 @@
                          '    git submodule update --init --recursive ext/fmt\n')
 
 # Check for googletest and checkout submodule if needed
-if env['system_googletest'] in ('y', 'default'):
-    if conf.CheckCXXHeader('gtest/gtest.h', '""'):
-        env['system_googletest'] = True
-    elif env['system_googletest'] == 'y':
-        config_error('Expected system installation of Googletest, but it '
-                     'could not be found.')
-
-if env['system_googletest'] in ('n', 'default'):
-    env['system_googletest'] = False
-    if not os.path.exists('ext/googletest/include/gtest/gtest.h'):
-        if not os.path.exists('.git'):
-            config_error('Googletest is missing. Install source in ext/googletest.')
-
-        try:
-            code = subprocess.call(['git','submodule','update','--init',
-                                    '--recursive','ext/googletest'])
-        except Exception:
-            code = -1
-        if code:
-            config_error('Googletest not found and submodule checkout failed.\n'
-                         'Try manually checking out the submodule with:\n\n'
-                         '    git submodule update --init --recursive ext/googletest\n')
+if env['googletest'] in ('y', 'default'):
+    if env['system_googletest'] in ('y', 'default'):
+        if conf.CheckCXXHeader('gtest/gtest.h', '""'):
+            env['system_googletest'] = True
+        elif env['system_googletest'] == 'y':
+            config_error('Expected system installation of Googletest, but it '
+                         'could not be found.')
+
+    if env['system_googletest'] in ('n', 'default'):
+        env['system_googletest'] = False
+        if not os.path.exists('ext/googletest/include/gtest/gtest.h'):
+            if not os.path.exists('.git'):
+                config_error('Googletest is missing. Install source in ext/googletest.')
+
+            try:
+                code = subprocess.call(['git','submodule','update','--init',
+                                        '--recursive','ext/googletest'])
+            except Exception:
+                code = -1
+            if code:
+                config_error('Googletest not found and submodule checkout failed.\n'
+                             'Try manually checking out the submodule with:\n\n'
+                             '    git submodule update --init --recursive ext/googletest\n')
 
 # Check for Eigen and checkout submodule if needed
 if env['system_eigen'] in ('y', 'default'):
