diff -Naur old/cmake/BuildPkgConfigFile.cmake new/cmake/BuildPkgConfigFile.cmake
--- old/cmake/BuildPkgConfigFile.cmake
+++ new/cmake/BuildPkgConfigFile.cmake
@@ -1,6 +1,12 @@
 FUNCTION(BUILD_PKCONFIG FILE)
   SET(LIBDIR "\\$\\{libdir\\}")
   STRING(REPLACE "\\" ""  LIBDIR ${LIBDIR})
+
+  SET(LAPACK_BLAS_LIBS ${LAPACK_LIBRARIES})
+  LIST(APPEND LAPACK_BLAS_LIBS ${BLAS_LIBRARIES})
+  LIST(REMOVE_DUPLICATES LAPACK_BLAS_LIBS)
+  STRING(REPLACE ";" " " LAPACK_BLAS_LIBS "${LAPACK_BLAS_LIBS}")
+
   IF(MKL_FOUND)
     IF(MKL_SCALAPACK_FOUND)
       SET(SF_LIBDIR "-L${LIBDIR} -lscifor ${MKL_LIBRARIES}")
@@ -16,13 +22,13 @@
       IF(APPLE)
 	SET(SF_LIBDIR "-L${LIBDIR} -lscifor ${SCALAPACK_LIBRARIES} -llapack -lblas")
       ELSE()
-	SET(SF_LIBDIR "-L${LIBDIR} -lscifor ${SCALAPACK_LIBRARIES} ${LAPACK_LIBRARIES} ${BLAS_LIBRARIES}")
+	SET(SF_LIBDIR "-L${LIBDIR} -lscifor ${SCALAPACK_LIBRARIES} ${LAPACK_BLAS_LIBS}")
       ENDIF(APPLE)
     ELSE()
       IF(APPLE)
 	SET(SF_LIBDIR "-L${LIBDIR} -lscifor -llapack -lblas")
       ELSE()
-	SET(SF_LIBDIR "-L${LIBDIR} -lscifor ${LAPACK_LIBRARIES} ${BLAS_LIBRARIES}")
+	SET(SF_LIBDIR "-L${LIBDIR} -lscifor ${LAPACK_BLAS_LIBS}")
       ENDIF(APPLE)
     ENDIF(SCALAPACK_FOUND)
   ELSE()
