Index: src/plugins/codecompletion/classbrowser.cpp
===================================================================
--- a/src/plugins/codecompletion/classbrowser.cpp	(revision 12288)
+++ b/src/plugins/codecompletion/classbrowser.cpp	(revision 12289)
@@ -1183,7 +1183,7 @@
 }
 
 #ifndef CC_NO_COLLAPSE_ITEM
-void ClassBrowser:CollapseItem(CCTreeItem* item)
+void ClassBrowser::CollapseItem(CCTreeItem* item)
 {
     wxTreeItemId Id = GetId(item);
     if (Id.IsOk())
