Index: codeblocks-20.03/src/plugins/contrib/FortranProject/fortranproject.h
===================================================================
--- codeblocks-20.03/src/plugins/contrib/FortranProject/fortranproject.h	(revision 294)
+++ codeblocks-20.03/src/plugins/contrib/FortranProject/fortranproject.h	(revision 295)
@@ -87,6 +87,7 @@
         void OnProjectFileAdded(CodeBlocksEvent& event);
         void OnProjectEndAddFiles(CodeBlocksEvent& event);
         void OnProjectFileRemoved(CodeBlocksEvent& event);
+        void OnProjectEndRemoveFiles(CodeBlocksEvent& event);
         void OnCompilerStarted(CodeBlocksEvent& event);
         void OnCleanProjectStarted(CodeBlocksEvent& event);
         void OnCleanWorkspaceStarted(CodeBlocksEvent& event);
Index: codeblocks-20.03/src/plugins/contrib/FortranProject/fortranproject.cpp
===================================================================
--- codeblocks-20.03/src/plugins/contrib/FortranProject/fortranproject.cpp	(revision 294)
+++ codeblocks-20.03/src/plugins/contrib/FortranProject/fortranproject.cpp	(revision 295)
@@ -167,6 +167,7 @@
     pm->RegisterEventSink(cbEVT_PROJECT_FILE_ADDED, new cbEventFunctor<FortranProject, CodeBlocksEvent>(this, &FortranProject::OnProjectFileAdded));
     pm->RegisterEventSink(cbEVT_PROJECT_END_ADD_FILES, new cbEventFunctor<FortranProject, CodeBlocksEvent>(this, &FortranProject::OnProjectEndAddFiles));
     pm->RegisterEventSink(cbEVT_PROJECT_FILE_REMOVED, new cbEventFunctor<FortranProject, CodeBlocksEvent>(this, &FortranProject::OnProjectFileRemoved));
+    pm->RegisterEventSink(cbEVT_PROJECT_END_REMOVE_FILES, new cbEventFunctor<FortranProject, CodeBlocksEvent>(this, &FortranProject::OnProjectEndRemoveFiles));
     pm->RegisterEventSink(cbEVT_COMPILER_STARTED, new cbEventFunctor<FortranProject, CodeBlocksEvent>(this, &FortranProject::OnCompilerStarted));
     pm->RegisterEventSink(cbEVT_CLEAN_PROJECT_STARTED, new cbEventFunctor<FortranProject, CodeBlocksEvent>(this, &FortranProject::OnCleanProjectStarted));
     pm->RegisterEventSink(cbEVT_CLEAN_WORKSPACE_STARTED, new cbEventFunctor<FortranProject, CodeBlocksEvent>(this, &FortranProject::OnCleanWorkspaceStarted));
@@ -346,6 +347,14 @@
     if (IsAttached() && m_InitDone)
     {
         m_pNativeParser->RemoveFileFromParser(event.GetString());
+    }
+    event.Skip();
+}
+
+void FortranProject::OnProjectEndRemoveFiles(CodeBlocksEvent& event)
+{
+    if (IsAttached() && m_InitDone)
+    {
         m_pNativeParser->UpdateWorkspaceBrowser();
     }
     event.Skip();
