diff -Naur '--exclude=*.Plo' '--exclude=Makefile*' codeblocks-17.12_init/src/plugins/astyle/asstreamiterator.cpp codeblocks-17.12/src/plugins/astyle/asstreamiterator.cpp
--- codeblocks-17.12_init/src/plugins/astyle/asstreamiterator.cpp	2017-11-01 02:32:42.000000000 +0300
+++ codeblocks-17.12/src/plugins/astyle/asstreamiterator.cpp	2018-05-07 11:35:03.000000000 +0300
@@ -29,6 +29,9 @@
 {
 }
 
+/* Stub function getPeekStart() to provide the compatibility of codeblocks-17.12 with dev-util/astyle-3.1 */
+std::streamoff ASStreamIterator::getPeekStart() const { return 0; }
+
 bool ASStreamIterator::hasMoreLines() const
 {
     return (*m_CharPtr) != 0;
diff -Naur '--exclude=*.Plo' '--exclude=Makefile*' codeblocks-17.12_init/src/plugins/astyle/asstreamiterator.h codeblocks-17.12/src/plugins/astyle/asstreamiterator.h
--- codeblocks-17.12_init/src/plugins/astyle/asstreamiterator.h	2017-11-01 02:32:42.000000000 +0300
+++ codeblocks-17.12/src/plugins/astyle/asstreamiterator.h	2018-05-07 10:50:21.000000000 +0300
@@ -20,6 +20,8 @@
         ASStreamIterator(cbEditor* cbe, const wxChar* in);
         virtual ~ASStreamIterator();
 
+        /* Stub function getPeekStart() to provide the compatibility of codeblocks-17.12 with dev-util/astyle-3.1 */
+        std::streamoff getPeekStart() const;
         bool hasMoreLines() const;
         int  getStreamLength() const;
         std::string nextLine(bool emptyLineWasDeleted = false);
diff -Naur '--exclude=*.Plo' '--exclude=Makefile*' codeblocks-17.12_init/src/plugins/astyle/dlgformattersettings.cpp codeblocks-17.12/src/plugins/astyle/dlgformattersettings.cpp
--- codeblocks-17.12_init/src/plugins/astyle/dlgformattersettings.cpp	2017-11-01 02:32:42.000000000 +0300
+++ codeblocks-17.12/src/plugins/astyle/dlgformattersettings.cpp	2018-05-07 10:09:54.000000000 +0300
@@ -40,7 +40,8 @@
   else if (XRCCTRL(*m_dlg, "rbVTK", wxRadioButton)->GetValue())
     formatter.setFormattingStyle(astyle::STYLE_VTK);
   else if (XRCCTRL(*m_dlg, "rbBanner", wxRadioButton)->GetValue())
-    formatter.setFormattingStyle(astyle::STYLE_BANNER);
+    /* STYLE_BANNER from dev-util/astyle-3.0.1 was renamed to STYLE_RATLIFF in dev-util/astyle-3.1 */
+    formatter.setFormattingStyle(astyle::STYLE_RATLIFF);
   else if (XRCCTRL(*m_dlg, "rbGNU", wxRadioButton)->GetValue())
     formatter.setFormattingStyle(astyle::STYLE_GNU);
   else if (XRCCTRL(*m_dlg, "rbLinux", wxRadioButton)->GetValue())
diff -Naur '--exclude=*.Plo' '--exclude=Makefile*' codeblocks-17.12_init/src/plugins/astyle/formattersettings.cpp codeblocks-17.12/src/plugins/astyle/formattersettings.cpp
--- codeblocks-17.12_init/src/plugins/astyle/formattersettings.cpp	2017-11-01 02:32:42.000000000 +0300
+++ codeblocks-17.12/src/plugins/astyle/formattersettings.cpp	2018-05-07 10:52:01.000000000 +0300
@@ -55,7 +55,8 @@
       break;
 
     case aspsBanner: // Banner
-      formatter.setFormattingStyle(astyle::STYLE_BANNER);
+      /* STYLE_BANNER from dev-util/astyle-3.0.1 was renamed to STYLE_RATLIFF in dev-util/astyle-3.1 */
+      formatter.setFormattingStyle(astyle::STYLE_RATLIFF);
       break;
 
     case aspsGnu: // GNU
