Index: codeblocks-20.03/src/plugins/contrib/FortranProject/fortranproject.cpp
===================================================================
--- codeblocks-20.03/src/plugins/contrib/FortranProject/fortranproject.cpp	(revision 295)
+++ codeblocks-20.03/src/plugins/contrib/FortranProject/fortranproject.cpp	(revision 296)
@@ -1579,11 +1579,17 @@
         m_pNativeParser->GetCallTipHighlight(definition, commasPos, hlStart, hlEnd);
     }
     else if (!definition.IsEmpty() && isUnique &&
-        (!isAfterPercent || ( isAfterPercent && result->GetCount() >= 2 && (result->Item(0)->m_TokenKind == tkProcedure) )))
+        (!isAfterPercent || ( isAfterPercent && result->GetCount() >= 2 &&
+                             (result->Item(0)->m_TokenKind == tkProcedure || result->Item(1)->m_TokenKind == tkProcedure) )))
     {
         m_pNativeParser->GetCallTipHighlight(definition, commasPos, hlStart, hlEnd);
         if (isAfterPercent)
-            token = result->Item(1);
+        {
+            if (result->Item(0)->m_TokenKind == tkProcedure)
+                token = result->Item(1); // Item(0)--tkProcedure; Item(1)--(tkSubroutine or tkFunction)
+            else if (result->Item(1)->m_TokenKind == tkProcedure && result->GetCount() >= 3)
+                token = result->Item(2); // Item(0)--tkInterface; Item(1)--tkProcedure; Item(2)--(tkSubroutine or tkFunction)
+        }
 
         if (token->m_TokenKind == tkSubroutine || token->m_TokenKind == tkFunction || token->m_TokenKind == tkType)
         {
