diff -Naur old/src/plugins/contrib/FortranProject/autoinsert.cpp new/src/plugins/contrib/FortranProject/autoinsert.cpp
--- old/src/plugins/contrib/FortranProject/autoinsert.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/autoinsert.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -1,8 +1,6 @@
 #include "autoinsert.h"
 
-#ifndef CB_PRECOMP
-    #include <configmanager.h>
-#endif
+#include <configmanager.h>
 #include <algorithm>
 
 #include "fortranfileext.h"
diff -Naur old/src/plugins/contrib/FortranProject/autoinsert.h new/src/plugins/contrib/FortranProject/autoinsert.h
--- old/src/plugins/contrib/FortranProject/autoinsert.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/autoinsert.h	2020-03-31 18:20:10.000000000 +0300
@@ -5,10 +5,9 @@
 #ifndef CB_PRECOMP
     #include <wx/string.h>
     #include <wx/regex.h>
-
-    #include <cbeditor.h>
-    #include <cbstyledtextctrl.h>
 #endif
+#include <cbeditor.h>
+#include <cbstyledtextctrl.h>
 #include <map>
 
 
diff -Naur old/src/plugins/contrib/FortranProject/bindto.cpp new/src/plugins/contrib/FortranProject/bindto.cpp
--- old/src/plugins/contrib/FortranProject/bindto.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/bindto.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -7,19 +7,19 @@
 
 #ifndef CB_PRECOMP
     #include <wx/regex.h>
-    #include <wx/tokenzr.h>
     #include <wx/textdlg.h>
     #include <wx/dirdlg.h>
     #include <wx/msgdlg.h>
 
-    #include <editormanager.h>
-    #include <cbstyledtextctrl.h>
-    #include <configmanager.h>
     #include <projectmanager.h>
     #include <logmanager.h>
     #include <cbproject.h>
     #include <compilerfactory.h>
 #endif
+#include <wx/tokenzr.h>
+#include <editormanager.h>
+#include <cbstyledtextctrl.h>
+#include <configmanager.h>
 
 //(*IdInit(Bindto)
 const long Bindto::ID_BTOACTIVEPROJECT = wxNewId();
diff -Naur old/src/plugins/contrib/FortranProject/bufferparserthread.h new/src/plugins/contrib/FortranProject/bufferparserthread.h
--- old/src/plugins/contrib/FortranProject/bufferparserthread.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/bufferparserthread.h	2020-03-31 18:20:10.000000000 +0300
@@ -6,9 +6,8 @@
 #ifndef CB_PRECOMP
     #include <wx/thread.h>
     #include <wx/event.h>
-
-    #include <cbthreadpool.h>
 #endif
+#include <cbthreadpool.h>
 
 class NativeParserF;
 
diff -Naur old/src/plugins/contrib/FortranProject/calltreeview.cpp new/src/plugins/contrib/FortranProject/calltreeview.cpp
--- old/src/plugins/contrib/FortranProject/calltreeview.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/calltreeview.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -8,11 +8,11 @@
 
     #include <sdk_events.h>
     #include <manager.h>
-    #include <editormanager.h>
     #include <logmanager.h>
-    #include <configmanager.h>
-    #include <cbeditor.h>
 #endif
+#include <editormanager.h>
+#include <configmanager.h>
+#include <cbeditor.h>
 #include <cmath>
 
 #include "fortranproject.h"
diff -Naur old/src/plugins/contrib/FortranProject/calltreeview.h new/src/plugins/contrib/FortranProject/calltreeview.h
--- old/src/plugins/contrib/FortranProject/calltreeview.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/calltreeview.h	2020-03-31 18:20:10.000000000 +0300
@@ -5,9 +5,8 @@
 #ifndef CB_PRECOMP
     #include <wx/panel.h>
     #include <wx/treectrl.h>
-
-    #include <cbthreadpool.h>
 #endif
+#include <cbthreadpool.h>
 
 #include "tokenf.h"
 #include "fpimagelist.h"
diff -Naur old/src/plugins/contrib/FortranProject/ccsmartfilter.h new/src/plugins/contrib/FortranProject/ccsmartfilter.h
--- old/src/plugins/contrib/FortranProject/ccsmartfilter.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/ccsmartfilter.h	2020-03-31 18:20:10.000000000 +0300
@@ -14,9 +14,8 @@
     #include <wx/string.h>
     #include <wx/event.h>
     #include <wx/file.h>
-
-    #include <cbeditor.h>
 #endif
+#include <cbeditor.h>
 #include <set>
 #include <vector>
 
diff -Naur old/src/plugins/contrib/FortranProject/changecase.cpp new/src/plugins/contrib/FortranProject/changecase.cpp
--- old/src/plugins/contrib/FortranProject/changecase.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/changecase.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -2,18 +2,17 @@
 #include "changecase.h"
 
 #ifndef CB_PRECOMP
-    #include <wx/tokenzr.h>
-
     #include <cbauibook.h>
-    #include <cbeditor.h>
     #include <cbproject.h>
-    #include <cbstyledtextctrl.h>
-    #include <editormanager.h>
     #include <editorcolourset.h>
     #include <logmanager.h>
     #include <projectmanager.h>
     #include <projectfile.h>
 #endif
+#include <wx/tokenzr.h>
+#include <cbeditor.h>
+#include <cbstyledtextctrl.h>
+#include <editormanager.h>
 #include <set>
 
 #include "fortranfileext.h"
diff -Naur old/src/plugins/contrib/FortranProject/constrhighlighter.cpp new/src/plugins/contrib/FortranProject/constrhighlighter.cpp
--- old/src/plugins/contrib/FortranProject/constrhighlighter.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/constrhighlighter.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -10,12 +10,12 @@
 
 #include <sdk.h>
 #ifndef CB_PRECOMP
-    #include <configmanager.h>
-    #include <editormanager.h>
     #include <logmanager.h>
-    #include <cbstyledtextctrl.h>
 #endif
 #include <algorithm>
+#include <configmanager.h>
+#include <editormanager.h>
+#include <cbstyledtextctrl.h>
 
 ConstrHighlighter::ConstrHighlighter():
     m_MakeHighlight(true),
diff -Naur old/src/plugins/contrib/FortranProject/constrhighlighter.h new/src/plugins/contrib/FortranProject/constrhighlighter.h
--- old/src/plugins/contrib/FortranProject/constrhighlighter.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/constrhighlighter.h	2020-03-31 18:20:10.000000000 +0300
@@ -8,9 +8,8 @@
     #include <wx/string.h>
     #include <wx/stopwatch.h>
     #include <wx/colour.h>
-
-    #include <cbeditor.h>
 #endif
+#include <cbeditor.h>
 #include <set>
 #include <list>
 #include <map>
diff -Naur old/src/plugins/contrib/FortranProject/docblock.cpp new/src/plugins/contrib/FortranProject/docblock.cpp
--- old/src/plugins/contrib/FortranProject/docblock.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/docblock.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -1,9 +1,5 @@
-
 #include "docblock.h"
-
-#ifndef CB_PRECOMP
-    #include "cbcolourmanager.h"
-#endif
+#include "cbcolourmanager.h"
 #include <iostream>
 
 DocBlock::DocBlock():
diff -Naur old/src/plugins/contrib/FortranProject/docblock.h new/src/plugins/contrib/FortranProject/docblock.h
--- old/src/plugins/contrib/FortranProject/docblock.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/docblock.h	2020-03-31 18:20:10.000000000 +0300
@@ -4,8 +4,8 @@
 #include <sdk.h>
 #ifndef CB_PRECOMP
     #include <wx/string.h>
-    #include <wx/html/htmlwin.h>
 #endif
+#include <wx/html/htmlwin.h>
 #include <map>
 
 #include "tokenf.h"
diff -Naur old/src/plugins/contrib/FortranProject/finfowindow.cpp new/src/plugins/contrib/FortranProject/finfowindow.cpp
--- old/src/plugins/contrib/FortranProject/finfowindow.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/finfowindow.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -10,12 +10,12 @@
 
 #include <sdk.h>
 #ifndef CB_PRECOMP
-    #include <configmanager.h>
     #include <manager.h>
     #include <editorcolourset.h>
-    #include <cbeditor.h>
     #include <wx/sizer.h>
 #endif
+#include <configmanager.h>
+#include <cbeditor.h>
 
 FInfoWindow::FInfoWindow()
     :wxPanel(Manager::Get()->GetAppWindow())
diff -Naur old/src/plugins/contrib/FortranProject/finfowindow.h new/src/plugins/contrib/FortranProject/finfowindow.h
--- old/src/plugins/contrib/FortranProject/finfowindow.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/finfowindow.h	2020-03-31 18:20:10.000000000 +0300
@@ -13,9 +13,8 @@
 #include <sdk.h>
 #ifndef CB_PRECOMP
     #include <wx/panel.h>
-    #include <cbstyledtextctrl.h>
 #endif
-
+#include <cbstyledtextctrl.h>
 
 class FInfoWindow : public wxPanel
 {
diff -Naur old/src/plugins/contrib/FortranProject/formatindent.cpp new/src/plugins/contrib/FortranProject/formatindent.cpp
--- old/src/plugins/contrib/FortranProject/formatindent.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/formatindent.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -2,14 +2,13 @@
 #include "formatindent.h"
 
 #ifndef CB_PRECOMP
-    #include <wx/tokenzr.h>
-
     #include <projectmanager.h>
     #include <cbproject.h>
-    #include <editormanager.h>
-    #include <configmanager.h>
     #include "cbeditor.h"
 #endif
+#include <wx/tokenzr.h>
+#include <editormanager.h>
+#include <configmanager.h>
 #include <vector>
 
 #include "formatindentdlg.h"
diff -Naur old/src/plugins/contrib/FortranProject/formatindent.h new/src/plugins/contrib/FortranProject/formatindent.h
--- old/src/plugins/contrib/FortranProject/formatindent.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/formatindent.h	2020-03-31 18:20:10.000000000 +0300
@@ -3,9 +3,7 @@
 #define INDENTMAKER_H
 
 #include <sdk.h>
-#ifndef CB_PRECOMP
-    #include <cbstyledtextctrl.h>
-#endif
+#include <cbstyledtextctrl.h>
 
 #include "indentestimator.h"
 
diff -Naur old/src/plugins/contrib/FortranProject/formatindentdlg.cpp new/src/plugins/contrib/FortranProject/formatindentdlg.cpp
--- old/src/plugins/contrib/FortranProject/formatindentdlg.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/formatindentdlg.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -1,13 +1,8 @@
-
 #include "formatindentdlg.h"
-
-#ifndef CB_PRECOMP
-    #include <editormanager.h>
-    #include <configmanager.h>
-    #include <cbstyledtextctrl.h>
-    #include "cbeditor.h"
-#endif
-
+#include <editormanager.h>
+#include <configmanager.h>
+#include <cbstyledtextctrl.h>
+#include "cbeditor.h"
 //(*InternalHeaders(FormatIndentDlg)
 #include <wx/intl.h>
 #include <wx/button.h>
diff -Naur old/src/plugins/contrib/FortranProject/fortranfileext.cpp new/src/plugins/contrib/FortranProject/fortranfileext.cpp
--- old/src/plugins/contrib/FortranProject/fortranfileext.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/fortranfileext.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -2,11 +2,11 @@
 #include "fortranfileext.h"
 
 #ifndef CB_PRECOMP
-    #include <wx/tokenzr.h>
-    #include <configmanager.h>
-    #include <editormanager.h>
     #include <logmanager.h>
 #endif
+#include <wx/tokenzr.h>
+#include <configmanager.h>
+#include <editormanager.h>
 
 FortranFileExt::FortranFileExt()
 {
diff -Naur old/src/plugins/contrib/FortranProject/fortranproject.cpp new/src/plugins/contrib/FortranProject/fortranproject.cpp
--- old/src/plugins/contrib/FortranProject/fortranproject.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/fortranproject.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -10,7 +10,6 @@
 #include <sdk.h> // Code::Blocks SDK
 #ifndef CB_PRECOMP
     #include <wx/filename.h>
-    #include <wx/tokenzr.h>
     #include <wx/gdicmn.h>
     #include <wx/xrc/xmlres.h>
     #include <wx/event.h>
@@ -18,18 +17,19 @@
     #include <wx/toolbar.h>
     #include <wx/choicdlg.h>
 
-    #include <configurationpanel.h>
     #include <manager.h>
-    #include <ccmanager.h>
     #include <editorcolourset.h>
-    #include <editormanager.h>
     #include <logmanager.h>
     #include <projectmanager.h>
-    #include <cbstyledtextctrl.h>
-    #include <projectloader_hooks.h>
-    #include <editor_hooks.h>
-    #include <cbeditor.h>
 #endif
+#include <wx/tokenzr.h>
+#include <configurationpanel.h>
+#include <ccmanager.h>
+#include <editormanager.h>
+#include <cbstyledtextctrl.h>
+#include <projectloader_hooks.h>
+#include <editor_hooks.h>
+#include <cbeditor.h>
 #include <vector>
 
 #include "fortranproject.h"
diff -Naur old/src/plugins/contrib/FortranProject/fpimagelist.cpp new/src/plugins/contrib/FortranProject/fpimagelist.cpp
--- old/src/plugins/contrib/FortranProject/fpimagelist.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/fpimagelist.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -4,9 +4,8 @@
 #ifndef CB_PRECOMP
     #include <wx/string.h>
     #include <wx/bitmap.h>
-
-    #include <configmanager.h>
 #endif
+#include <configmanager.h>
 
 FPImageList::FPImageList(int imSize)
 {
diff -Naur old/src/plugins/contrib/FortranProject/fpoptionsdlg.cpp new/src/plugins/contrib/FortranProject/fpoptionsdlg.cpp
--- old/src/plugins/contrib/FortranProject/fpoptionsdlg.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/fpoptionsdlg.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -22,11 +22,11 @@
     #include <wx/colordlg.h>
     #include <wx/choice.h>
 
-    #include <configmanager.h>
     #include <manager.h>
     #include <globals.h>
 #endif
 #include <algorithm>
+#include <configmanager.h>
 #include <vector>
 
 #include "fortranproject.h"
diff -Naur old/src/plugins/contrib/FortranProject/fpoptionsdlg.h new/src/plugins/contrib/FortranProject/fpoptionsdlg.h
--- old/src/plugins/contrib/FortranProject/fpoptionsdlg.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/fpoptionsdlg.h	2020-03-31 18:20:10.000000000 +0300
@@ -10,9 +10,9 @@
 #ifndef CB_PRECOMP
     #include <wx/intl.h>
 
-    #include <configurationpanel.h>
     #include <settings.h>
 #endif
+#include <configurationpanel.h>
 
 #include "fortranproject.h"
 #include "workspacebrowserf.h"
diff -Naur old/src/plugins/contrib/FortranProject/fpoptionsprjdlg.cpp new/src/plugins/contrib/FortranProject/fpoptionsprjdlg.cpp
--- old/src/plugins/contrib/FortranProject/fpoptionsprjdlg.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/fpoptionsprjdlg.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -14,12 +14,13 @@
     #include <wx/xrc/xmlres.h>
 
     #include <cbproject.h>
-    #include <cbstyledtextctrl.h>
+
     #include <globals.h>
     #include <logmanager.h>
     #include <manager.h>
 #endif
 
+#include <cbstyledtextctrl.h>
 #include <editpathdlg.h>
 
 BEGIN_EVENT_TABLE(FPOptionsProjectDlg, wxPanel)
diff -Naur old/src/plugins/contrib/FortranProject/fpoptionsprjdlg.h new/src/plugins/contrib/FortranProject/fpoptionsprjdlg.h
--- old/src/plugins/contrib/FortranProject/fpoptionsprjdlg.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/fpoptionsprjdlg.h	2020-03-31 18:20:10.000000000 +0300
@@ -10,8 +10,8 @@
 #ifndef CB_PRECOMP
     #include <wx/intl.h>
     #include <settings.h>
-    #include <configurationpanel.h>
 #endif
+#include <configurationpanel.h>
 
 #include "nativeparserf.h"
 #include "parserf.h"
diff -Naur old/src/plugins/contrib/FortranProject/indentestimator.cpp new/src/plugins/contrib/FortranProject/indentestimator.cpp
--- old/src/plugins/contrib/FortranProject/indentestimator.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/indentestimator.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -1,11 +1,9 @@
 
 #include "indentestimator.h"
-
 #include <sdk.h>
-#ifndef CB_PRECOMP
-    #include <wx/tokenzr.h>
-    #include <configmanager.h>
-#endif
+#include <wx/tokenzr.h>
+
+#include <configmanager.h>
 
 FormatIndentCodeTree::FormatIndentCodeTree()
 {
diff -Naur old/src/plugins/contrib/FortranProject/keywordsparserf.cpp new/src/plugins/contrib/FortranProject/keywordsparserf.cpp
--- old/src/plugins/contrib/FortranProject/keywordsparserf.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/keywordsparserf.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -8,16 +8,16 @@
 
 #include <sdk.h>
 #ifndef CB_PRECOMP
-    #include <wx/tokenzr.h>
     #include <wx/string.h>
     #include <wx/arrstr.h>
     #include <wx/file.h>
 
     #include <globals.h>
-    #include <cbstyledtextctrl.h>
-    #include <configmanager.h>
     #include <logmanager.h>
 #endif
+#include <wx/tokenzr.h>
+#include <cbstyledtextctrl.h>
+#include <configmanager.h>
 
 #include "parserthreadf.h"
 
diff -Naur old/src/plugins/contrib/FortranProject/keywordsparserf.h new/src/plugins/contrib/FortranProject/keywordsparserf.h
--- old/src/plugins/contrib/FortranProject/keywordsparserf.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/keywordsparserf.h	2020-03-31 18:20:10.000000000 +0300
@@ -12,9 +12,8 @@
     #include <wx/string.h>
     #include <wx/event.h>
     #include <wx/file.h>
-
-    #include <cbeditor.h>
 #endif
+#include <cbeditor.h>
 #include <set>
 #include <vector>
 
diff -Naur old/src/plugins/contrib/FortranProject/nativeparserf.cpp new/src/plugins/contrib/FortranProject/nativeparserf.cpp
--- old/src/plugins/contrib/FortranProject/nativeparserf.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/nativeparserf.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -11,25 +11,25 @@
     #include <wx/regex.h>
     #include <wx/log.h>
     #include <wx/string.h>
-    #include <wx/tokenzr.h>
     #include <wx/dir.h>
     #include <wx/wfstream.h>
     #include <wx/stopwatch.h>
 
     #include <manager.h>
-    #include <configmanager.h>
-    #include <editormanager.h>
     #include <projectmanager.h>
     #include <pluginmanager.h>
     #include <logmanager.h>
     #include <cbauibook.h>
-    #include <cbeditor.h>
     #include <cbproject.h>
     #include <cbexception.h>
-    #include <projectloader_hooks.h>
-    #include <cbstyledtextctrl.h>
     #include <tinyxml.h>
 #endif
+#include <wx/tokenzr.h>
+#include <configmanager.h>
+#include <editormanager.h>
+#include <cbeditor.h>
+#include <projectloader_hooks.h>
+#include <cbstyledtextctrl.h>
 #include <cctype>
 
 #include "workspacebrowserf.h"
diff -Naur old/src/plugins/contrib/FortranProject/nativeparserf.h new/src/plugins/contrib/FortranProject/nativeparserf.h
--- old/src/plugins/contrib/FortranProject/nativeparserf.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/nativeparserf.h	2020-03-31 18:20:10.000000000 +0300
@@ -10,9 +10,8 @@
 #ifndef CB_PRECOMP
     #include <wx/string.h>
     #include <wx/event.h>
-
-    #include <cbthreadpool.h>
 #endif
+#include <cbthreadpool.h>
 #include <map>
 #include <set>
 
diff -Naur old/src/plugins/contrib/FortranProject/parserf.cpp new/src/plugins/contrib/FortranProject/parserf.cpp
--- old/src/plugins/contrib/FortranProject/parserf.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/parserf.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -8,18 +8,18 @@
 
 #include <sdk.h>
 #ifndef CB_PRECOMP
-    #include <wx/tokenzr.h>
     #include <wx/string.h>
     #include <wx/thread.h>
     #include <wx/arrstr.h>
     #include <wx/regex.h>
 
-    #include <cbstyledtextctrl.h>
-    #include <configmanager.h>
-    #include <editormanager.h>
     #include <globals.h>
     #include <logmanager.h>
 #endif
+#include <wx/tokenzr.h>
+#include <cbstyledtextctrl.h>
+#include <configmanager.h>
+#include <editormanager.h>
 #include <vector>
 
 #include "workspaceparserthread.h"
diff -Naur old/src/plugins/contrib/FortranProject/parserf.h new/src/plugins/contrib/FortranProject/parserf.h
--- old/src/plugins/contrib/FortranProject/parserf.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/parserf.h	2020-03-31 18:20:10.000000000 +0300
@@ -12,9 +12,8 @@
     #include <wx/string.h>
     #include <wx/event.h>
     #include <wx/file.h>
-
-    #include <cbeditor.h>
 #endif
+#include <cbeditor.h>
 #include <set>
 #include <vector>
 
diff -Naur old/src/plugins/contrib/FortranProject/parserthreadf.cpp new/src/plugins/contrib/FortranProject/parserthreadf.cpp
--- old/src/plugins/contrib/FortranProject/parserthreadf.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/parserthreadf.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -7,9 +7,7 @@
 #include "parserthreadf.h"
 
 #include <sdk.h>
-#ifndef CB_PRECOMP
-    #include <wx/tokenzr.h>
-#endif
+#include <wx/tokenzr.h>
 #include <set>
 
 #include "usetokenf.h"
diff -Naur old/src/plugins/contrib/FortranProject/tab2space.cpp new/src/plugins/contrib/FortranProject/tab2space.cpp
--- old/src/plugins/contrib/FortranProject/tab2space.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/tab2space.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -3,15 +3,14 @@
 
 #include <sdk.h>
 #ifndef CB_PRECOMP
-    #include <wx/tokenzr.h>
-
-    #include <cbeditor.h>
     #include <cbproject.h>
-    #include <cbstyledtextctrl.h>
-    #include <editormanager.h>
     #include <logmanager.h>
     #include <projectmanager.h>
 #endif
+#include <wx/tokenzr.h>
+#include <cbeditor.h>
+#include <cbstyledtextctrl.h>
+#include <editormanager.h>
 #include <set>
 
 #include "fortranfileext.h"
diff -Naur old/src/plugins/contrib/FortranProject/workspacebrowserf.cpp new/src/plugins/contrib/FortranProject/workspacebrowserf.cpp
--- old/src/plugins/contrib/FortranProject/workspacebrowserf.cpp	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/workspacebrowserf.cpp	2020-03-31 18:20:10.000000000 +0300
@@ -19,21 +19,21 @@
     #include <wx/button.h>
     #include <wx/utils.h> // wxBusyCursor
     #include <wx/tipwin.h>
-    #include <wx/tokenzr.h>
     #include <wx/combobox.h>
     #include <wx/settings.h>
     #include <wx/choicdlg.h>
 
     #include <cbproject.h>
-    #include <cbeditor.h>
-    #include <configmanager.h>
-    #include <editormanager.h>
     #include <globals.h>
     #include <manager.h>
     #include <pluginmanager.h>
     #include <projectmanager.h>
-    #include <cbstyledtextctrl.h>
 #endif
+#include <wx/tokenzr.h>
+#include <cbeditor.h>
+#include <configmanager.h>
+#include <editormanager.h>
+#include <cbstyledtextctrl.h>
 
 #include "nativeparserf.h"
 #include "fortranproject.h"
diff -Naur old/src/plugins/contrib/FortranProject/workspaceparserthread.h new/src/plugins/contrib/FortranProject/workspaceparserthread.h
--- old/src/plugins/contrib/FortranProject/workspaceparserthread.h	2020-03-28 15:36:11.000000000 +0300
+++ new/src/plugins/contrib/FortranProject/workspaceparserthread.h	2020-03-31 18:20:10.000000000 +0300
@@ -5,10 +5,8 @@
 #ifndef CB_PRECOMP
     #include <wx/thread.h>
     #include <wx/event.h>
-
-    #include <cbthreadpool.h>
 #endif
-
+#include <cbthreadpool.h>
 
 extern wxMutex s_WorkspaceParserMutex;
 extern wxMutex s_NewTokensMutex;
